[Feature] Add Splash screen

S-H-Y-A opened the pull request:
<!--- Please fill the necessary details below -->
## Purpose / Description
Splash screen is one of new Android features and it makes user experiences richer.

This animation is Inspired by the navbar branding image of [Anki Web](https://ankiweb.net/)

## Fixes
* Fixes #14838

## Approach
create splash screen

| ~android 11 | android 12 |
| ---- | ---- |
| ![Screenshot_20231128_223536](https://github.com/ankidroid/Anki-Android/assets/74596628/b1c898e8-8535-43c1-a6a0-95fd9a6f6629) | <video src="https://github.com/ankidroid/Anki-Android/assets/74596628/eec5a035-d028-4d06-83b7-dfafa9c6b724"></video> |

## How Has This Been Tested?
- physical device (Android 13)
- emulator on Android Studio (Android 7.1)
P.S the difference of color of star is already fixed.

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

criticalAY commented:
Is this really needed that too using a dependency? we already have a splash screen ig

S-H-Y-A commented:
> Is this really needed that too using a dependency? we already have a splash screen ig

Ops, this fallback method for android11 and below seems not to depend on androidx.core.core-splashscreen.
I'll remove the dependency within a few hours (because I'm not in home now.)

david-allison reviewed (dismissed):
I love this

Lint is failing, can't immediately see why, could you take a look?

From the video, the spinning animation seems much too fast, I'd see about reducing the speed to maybe 0.75x what's currently there.

S-H-Y-A commented:
I can't determine what is wrong because the lint error happens in the code whch has'nt been modified.
I'll try to reset my git commits and recommit.

david-allison commented:
Thanks for trying, I'll take a look

S-H-Y-A commented:
> From the video, the spinning animation seems much too fast, I'd see about reducing the speed to maybe 0.75x what's currently there.

Thank you for your advice.
I changed the rotation speed, however, both my smartphone and pc are not high end models, and it was found I couldn't record the screen as usual. So, I would like to ask @david-allison to check the animation yourself.

david-allison commented on AnkiDroid/src/main/res/drawable/splash_icon.xml:
```diff
@@ -0,0 +1,353 @@
+                <objectAnimator
+                    android:propertyName="rotation"
+                    android:startOffset="2500"
+                    android:duration="2000"
```
Personal preference

```suggestion
                    android:duration="3000"
```

S-H-Y-A replied:
I changed the duration on [this commit](https://github.com/ankidroid/Anki-Android/pull/14837/commits/18349689f28905925c416ed9f4e4597d5b1ed511).

david-allison reviewed (changes requested):
On my phone/emulator, the opposite problem occurs:

There's not enough time to show the 'star' before the splash screen disappears

When I add a sleep, the star spins too quickly

criticalAY commented on AnkiDroid/src/main/res/drawable/splash_icon.xml:
```diff
@@ -0,0 +1,353 @@
+        <aapt:attr name="android:animation">
+            <set>
+<!--                Behavior between physical devices and Android Studio preview are different. -->
+<!--                <objectAnimator
```
I guess this commented out code is not required?

S-H-Y-A replied:
The animation is correct without The commented out code on physical device, but it is needed for Preview of Android Studio Preview.

BrayanDSO commented:
> There's not enough time to show the 'star' before the splash screen disappears

Same here. In those cases, having no animation looks better IMHO

S-H-Y-A commented:
> On my phone/emulator, the opposite problem occurs:
> 
> There's not enough time to show the 'star' before the splash screen disappears
> 
> When I add a sleep, the star spins too quickly
> 
> Don't worry about the lint errors for now, I've added this to make them easier to diagnose, should be in `main` later:
> 
> * [chore(lint): cleanup CI output on failure #14854](https://github.com/ankidroid/Anki-Android/pull/14854)

I fixed these problems, however, I realized a hack* to rotate star infinitely doesn't work, so the star will rotate jsut once.

I thought It might be better not to rotate the star.

<hr>
* This hack is to set 
```
                    android:repeatMode="restart"
                    android:repeatCount="infinite"
```
to `objectAnimator`.

Splash screen is removed when  `windowSplashScreenAnimationDuration` pass and the system is preparing to draw the view.

On Android 12 the hack works well because `windowSplashScreenAnimationDuration` is set in style.xml manually.
On Android 13 and avobe, however, the system set  `windowSplashScreenAnimationDuration`  `0` automatically and start removing splash screen as soon as it's ready and then splash screen doesn't last until the animation finish.
The reason why it happens is that the system on Android 11 and avobe tries to estimate the duration of the vector animation automatically, but it can't because of the hack.

Due to the bad performance of my smartphone, I couldn't realize that the hack doesn't work.

david-allison reviewed (commented):
I feel this would be cleaner as:

```kotlin

    /** @see executeFunctionAfterSplashScreen */
    override fun onCreate(savedInstanceState: Bundle?) {
        // Note: This is our entry point from the launcher with intent: android.intent.action.MAIN
        executeFunctionAfterSplashScreen(minDurationMs = 1000L) {
            Themes.setTheme(this)
            disableXiaomiForceDarkMode(this)
            val intent = intent
            Timber.v(intent.toString())
            val reloadIntent = Intent(this, DeckPicker::class.java)
            reloadIntent.setDataAndType(getIntent().data, getIntent().type)
            val action = intent.action
            // #6157 - We want to block actions that need permissions we don't have, but not the default case
            // as this requires nothing
            val runIfStoragePermissions = Consumer { runnable: Runnable ->
                performActionIfStorageAccessible(runnable, reloadIntent, action)
            }
            when (getLaunchType(intent)) {
                LaunchType.FILE_IMPORT -> runIfStoragePermissions.accept(Runnable {
                    handleFileImport(
                        intent,
                        reloadIntent,
                        action
                    )
                })

                LaunchType.SYNC -> runIfStoragePermissions.accept(Runnable {
                    handleSyncIntent(
                        reloadIntent,
                        action
                    )
                })

                LaunchType.REVIEW -> runIfStoragePermissions.accept(Runnable { handleReviewIntent(intent) })
                LaunchType.DEFAULT_START_APP_IF_NEW -> {
                    Timber.d("onCreate() performing default action")
                    launchDeckPickerIfNoOtherTasks(reloadIntent)
                }

                LaunchType.COPY_DEBUG_INFO -> {
                    copyDebugInfoToClipboard(intent)
                    finish()
                }
            }
        }
        super.onCreate(savedInstanceState)
    }

    private fun executeFunctionAfterSplashScreen(minDurationMs: Long, function: () -> Unit) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) {
            executeFunctionWithDelay(minDurationMs, function)
            return
        }
        splashScreen.setOnExitAnimationListener { splashScreenView ->
            val animationDuration = splashScreenView.iconAnimationDuration
            val animationStart = splashScreenView.iconAnimationStart

            Timber.v("duration :${animationDuration?.toMillis()} start $animationStart")
            val remainingDuration = when {
                animationDuration == null -> 0L
                animationStart == null -> 0L
                else -> (animationDuration - Duration.between(animationStart, Instant.ofEpochMilli(TimeManager.time.intTimeMS())))
                    .toMillis()
            }.coerceAtLeast(0L)
            Timber.v("remainDuration :$remainingDuration")
            executeFunctionWithDelay(remainingDuration, function)
        }
    }
```

+ simplifying `Duration.between(animationStart, Instant.ofEpochMilli(TimeManager.time.intTimeMS()))`

david-allison reviewed (changes requested):
Submitted the review too early (only tested on API 33):

The splash screen duration is too long 
* The 'appear' animation needs to be faster
* the star shouldn't need to spin before it closes

🪲 API 33:
* reopening AnkiDroid leads to a grey screen

criticalAY commented:
Would like to take over it

david-allison commented:
Sounds good

Haz3-jolt commented:
@criticalAY if you don't mind I've picked this one up
